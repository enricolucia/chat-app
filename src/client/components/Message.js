import React from 'react'
import Countdown from './Countdown'
import { darken } from 'polished'
import styled, { keyframes } from 'styled-components'

const yourBubblePrimary = '#329cd4'
const myBubblePrimary = '#AFFFC9'
const darkGrey = '#A9A9A9'

const slideInFromRight = keyframes`
  from {
    opacity: 0;
    transform: translateX(50%);
  }
  to {
    opacity: 1;
    transform: translateX(0%);
  }
`

const slideInFromLeft = keyframes`
  from {
    opacity: 0;
    transform: translateX(-50%);
  }
  to {
    opacity: 1;
    transform: translateX(0%);
  }
`

const StyledMessage = styled.div`
  padding: 8px;
  border-radius: 4px;
  margin: 10px;
  position: relative;
  border-style: solid;
  border-width: 1px 0 1px;
  max-width: 70%;
  animation: ${props => props.data.own
    ? slideInFromRight
    : slideInFromLeft
  } 0.2s linear;

  ${props => props.data.highlighted && `font-size: 110%;`}
  ${props => props.data.faded && `opacity: 0.1;`}
  ${props => props.data.think && `color: ${darkGrey};`}

  :before {
    content: '';
    position: absolute;
    display: block;
    z-index: 1;
    border-style: solid;
  }

  ${props => props.data.own && `
    background: ${props.data.highlighted
      ? `${darken(0.1, myBubblePrimary)}`
      : `${myBubblePrimary}`
    }
    margin-left: auto;
    margin-right: 10px;
    border: 1px solid ${darken(0.1, myBubblePrimary)};

    :before {
      border-width: 6px 0 6px 10px;
      border-color: transparent ${darken(0.1, myBubblePrimary)};
      right: -11px;
      bottom: 2px;
    }
  `}
  ${props => !props.data.own && `
    background: ${props.data.highlighted
      ? `${darken(0.1, yourBubblePrimary)}`
      : `${yourBubblePrimary}`
    }
    margin-right: auto;
    margin-left: 10px;
    border: 1px solid ${darken(0.1, yourBubblePrimary)};

    :before {
      border-width: 6px 10px 6px 0;
      border-color: transparent ${darken(0.1, yourBubblePrimary)};
      left: -11px;
      bottom: 2px;
    }
  `}
}`

const Message = ({ data, isCountdown }) => {
  return (
    <StyledMessage data={data}>
      {isCountdown
        ? <Countdown data={data} />
        : data.content
      }
    </StyledMessage>
  )
}

export default Message
