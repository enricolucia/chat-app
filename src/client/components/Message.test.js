/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'

import Message from './Message'

const defaultProps = {
  id: '0',
  data: {
    countdown: false,
    content: 'Lorem ipsum',
    highlighted: false,
    think: false,
    own: true
  }
}

describe('Message', () => {
  it('renders with a plain content', () => {
    const tree = renderer.create(
      <Message {...defaultProps} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('renders with a countdown content', () => {
    const newProps = Object.assign({}, defaultProps, {countdown: true})
    const tree = renderer.create(
      <Message {...newProps} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
