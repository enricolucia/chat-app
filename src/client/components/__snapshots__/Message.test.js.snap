// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message renders with a countdown content 1`] = `
<div
  className="Message__StyledMessage-ktMMqk dHPFVv"
  data={
    Object {
      "content": "Lorem ipsum",
      "countdown": false,
      "highlighted": false,
      "own": true,
      "think": false,
    }
  }
>
  Lorem ipsum
</div>
`;

exports[`Message renders with a plain content 1`] = `
<div
  className="Message__StyledMessage-ktMMqk dHPFVv"
  data={
    Object {
      "content": "Lorem ipsum",
      "countdown": false,
      "highlighted": false,
      "own": true,
      "think": false,
    }
  }
>
  Lorem ipsum
</div>
`;
